lib/mysql/mysql.go

```diff
@@ -0,0 +1,651 @@
+	STATE_SSL_HANDSHAKE = "SSL_HANDSHAKE"
+
+	// After connection has been negotiated (from either CONNECTED or SSL_HANDSHAKE)
+	STATE_FINISHED = "STATE_FINISHED"
+)
+
+type ConnectionState string
+
+const (
+	PACKET_TYPE_OK          = "OK"
+	PACKET_TYPE_ERROR       = "ERROR"
+	PACKET_TYPE_HANDSHAKE   = "HANDSHAKE"
+	PACKET_TYPE_EOF         = "EOF"
+	PACKET_TYPE_SSL_REQUEST = "SSL_REQUEST"
+)
+
+type PacketType string
+
+// Capability flags
+const (
+	CLIENT_LONG_PASSWORD uint32 = (1 << iota)
+	CLIENT_FOUND_ROWS
+	CLIENT_LONG_FLAG
+	CLIENT_CONNECT_WITH_DB
+	CLIENT_NO_SCHEMA
+	CLIENT_COMPRESS
+	CLIENT_ODBC
+	CLIENT_LOCAL_FILES
+	CLIENT_IGNORE_SPACE
+	CLIENT_PROTOCOL_41
+	CLIENT_INTERACTIVE
+	// CLIENT_SSL = (1<<11)
+	CLIENT_SSL
+	CLIENT_IGNORE_SIGPIPE
+	CLIENT_TRANSACTIONS
+	CLIENT_RESERVED
+	// CLIENT_SECURE_CONNECTION = (1 << 15)
+	CLIENT_SECURE_CONNECTION
+	CLIENT_MULTI_STATEMENTS
+	CLIENT_MULTI_RESULTS
+	CLIENT_PS_MULTI_RESULTS
+	// CLIENT_PLUGIN_AUTH = ( 1 << 19 )
+	CLIENT_PLUGIN_AUTH
+	CLIENT_CONNECT_ATTRS
+	CLIENT_PLUGIN_AUTH_LEN_ENC_CLIENT_DATA
+	CLIENT_CAN_HANDLE_EXPIRED_PASSWORDS
+	CLIENT_SESSION_TRACK
+	CLIENT_DEPRECATED_EOF
+)
+
+// Config defaults
+const (
+	DEFAULT_TIMEOUT_SECS        = 3
+	DEFAULT_PORT                = 3306
+	DEFAULT_CLIENT_CAPABILITIES = CLIENT_SSL
+	DEFAULT_RESERVED_DATA_HEX   = "0000000000000000000000000000000000000000000000"
+)
+
+type Config struct {
+	// @TODO: Does it make sense to make Host/Port connection fields, so that Config can be shared across connections?
+	Host string
+	Port uint16
+
+	TLSConfig          *tls.Config
+	Timeout            time.Duration
+	ClientCapabilities uint32
+	MaxPacketSize      uint32
+	CharSet            byte
+	ReservedData       []byte
+}
+
+// Fill in a (possibly newly-created) Config instance with the default values
+func NewConfig(base *Config) *Config {
+	if base == nil {
+		base = &Config{}
+	}
+	if base.TLSConfig == nil {
+		// @TODO @FIXME Can this be pulled from a global zgrab config module?
+		base.TLSConfig = &tls.Config{InsecureSkipVerify: true}
+	}
+	if base.Port == 0 {
+		base.Port = DEFAULT_PORT
+	}
+	if base.Timeout == 0 {
+		base.Timeout = DEFAULT_TIMEOUT_SECS * time.Second
+	}
+	if base.ClientCapabilities == 0 {
+		base.ClientCapabilities = DEFAULT_CLIENT_CAPABILITIES
+	}
+	if base.ReservedData == nil {
+		bin, err := hex.DecodeString(DEFAULT_RESERVED_DATA_HEX)
+		if err != nil {
+			log.Fatalf("Invalid constant")
+		}
+		base.ReservedData = bin
+	}
+	return base
+}
+
+// Struct holding state for a single connection
+type Connection struct {
+	// Configuration for this connection
+	Config *Config
+
+	// Enum to track connection status
+	State ConnectionState
+	// TCP or TLS-wrapped Connection pointer (IsSecure will tell which)
+	Connection *net.Conn
+	// The sequence number used with the server to number packets
+	SequenceNumber uint8
+	// The "Handshake" packet sent by the server, holding flags used in future calls
+	Handshake *Handshake_Packet
```

Comment:
There aren't normally underscores in types. Are you matching to an RFC or something here?

Reply:
Sort of — the OK_Packet/ERR_Packet had underscores in the reference manual, but Handshake did not. Since it’s already straying from the spec, it sounds like OKPacket, ERRPacket (or ErrPacket, or ErrorPacket), HandshakePacket would be just as consistent with the spec, and more consistent with Go